### PR TITLE
Introduce extension APIs to handle exceptions from lifecycle callback methods

### PR DESCRIPTION
**Pull request by JockX**
## Overview

This change is an alternative take on issue #1454 based on still not merged PR https://github.com/junit-team/junit5/pull/1509. In this version there are separate interfaces supporting exception handling in each of `@BeforeAll`, `@BeforeEach`, `@AfterEach` and `@AfterAll`, so that it is not forced on the handler to implement `handleTestExecutionException` in addition to `handle(Before/After)(/All/Each)ExecutionException`

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by marcphilipp** [comment]
@JockX Thanks for the PR! We'll discuss it in the next team meeting this Friday.

**Comment by marcphilipp** [comment]
**Team Decision:** Please put all of the new methods into a single `LifecycleMethodExecutionExceptionHandler` interface, rename the methods to `handle[...]MethodExecutionException` (add `Method`), make them default methods with empty implementations, and throw them into the [kitchen sink](https://github.com/junit-team/junit5/blob/master/junit-jupiter-engine/src/test/java/org/junit/jupiter/api/extension/KitchenSinkExtension.java).

**Comment by marcphilipp** [comment]
@JockX Thanks for updating the PR! Please run `./gradlew spotlessApply` to fix the build.

**Comment by JockX** [comment]
@marcphilipp I would appreciate some help regarding the failure in azure tests. Does not seem to be related to the changes introduced in this PR

**Comment by marcphilipp** [comment]
No, that's related to me merging #1885, I'm afraid. I'll try to fix it ASAP.

**Comment by marcphilipp** [comment]
@JockX It's fixed in `master` now. Please rebase once again.

File: documentation/src/docs/asciidoc/user-guide/extensions.adoc [hunk 1]
```diff
@@ -433,18 +433,51 @@ INFO: Method [sleep50ms] took 53 ms.
...
+exception was never thrown. Handlers may also choose to rethrow the exception or throw
+a different one, potentially wrapping the original.
+
+Extensions implementing `{LifecycleMethodExecutionExceptionHandler}` that wish to handle
+exceptions thrown during `@BeforeAll` or `@AfterAll` need to be registered on a class level,
+while handlers for `BeforeEach` and `AfterEach` may be also registered for individual
+test methods.
+
+[source,java,indent=0]
+.An exception handling extension that records application state on error
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/extensions.adoc]:
```suggestion
.Registering multiple exception handling extensions
```

File: documentation/src/test/java/example/exception/MultipleHandlersTestCase.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    static class ThirdExecutedHandler implements LifecycleMethodExecutionExceptionHandler {
+        @Override
+        public void handleBeforeAllMethodExecutionException(ExtensionContext context, Throwable ex)
+                throws Throwable {
+            throw ex;
+        }
+    }
+}
```
Comment (marcphilipp) [inline documentation/src/test/java/example/exception/MultipleHandlersTestCase.java]:
```suggestion
	// tag::user_guide[]
}
// end::user_guide[]
```

File: documentation/src/test/java/example/exception/MultipleHandlersTestCase.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.LifecycleMethodExecutionExceptionHandler;
+import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
+
+// @formatter:off
+// tag::user_guide[]
+// Register handlers for @Test, @BeforeEach, @AfterEach as well as @BeforeAll and @AfterAll
+@ExtendWith({ThirdExecutedHandler.class})
```
Comment (marcphilipp) [inline documentation/src/test/java/example/exception/MultipleHandlersTestCase.java]:
```suggestion
@ExtendWith(ThirdExecutedHandler.class)
```

File: documentation/src/test/java/example/exception/RecordStateOnErrorExtension.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            throws Throwable {
+        memoryDumpForFurtherInvestigation("Failure recorded during class cleanup");
+        throw ex;
+    }
+    // end::user_guide[]
+
+    private void memoryDumpForFurtherInvestigation(String error) {
+
+    }
+}
```
Comment (marcphilipp) [inline documentation/src/test/java/example/exception/RecordStateOnErrorExtension.java]:
```suggestion
	// tag::user_guide[]
}
// end::user_guide[]
```

**Comment by sbrannen** [comment]
Actually, I'll see if I can merge this PR and resolve the conflicts myself.

So, please hold off on making any further changes at this time.

**Comment by sbrannen** [comment]
This has been merged into `master` in 2844ee40948b9ae14bc7426ad73c3400dbfff3a5.

However, @JockX, it would be great if you could provide the source for the PNG you supplied.

Thanks in advance!

**Comment by JockX** [comment]
@sbrannen, the png is a screenshot of a word document I created to more or less mimic the original. Should I put it in the repo? How is this usually handled?

**Comment by sbrannen** [comment]
> @sbrannen, the png is a screenshot of a word document I created to more or less mimic the original. Should I put it in the repo? 

Yes, that would be great! 👍 

> How is this usually handled?

Funny you should ask... for this particular graphic, the original author lost the original source, meaning we've had no way to _edit_ it. So now that you have a "source", we would be very grateful to gain access to it. 😉

**Comment by sbrannen** [comment]
Note that this PR was further refined in 8aaac855bb36d3d92d769f4c4e8e2e47d1ac6fa6.

**Pull request closed without merging**